Rootless (#879)
rstudio/helm: charts/rstudio-workbench/NEWS.md

# Changelog

## 0.21.1

- Workbench pods can now run as a non-root user. Set `pod.runAsRoot: false` (new value, default `true`) to run unprivileged.

> - When running unprivelaged SCIM must be used for user management.

Reviewer: ```suggestion
- Use SCIM for user management when running unprivileged.
```

- A new `config.sssd` block controls the bundled SSSD daemon (replaces the deprecated `config.userProvisioning`).

> - `config.userProvisioning` is deprecated; use `config.sssd.conf`. A warning is emitted when it is set.

Reviewer: Consider rewriting these two lines into a single bullet point; they feel kind of redundant. I'm also confused by `config.sssd` vs `config.sssd.conf`.

> - The Workbench and launcher prestart scripts (`prestart-workbench.bash`, `prestart-launcher.bash`) no longer perform root-only operations.

Reviewer: How relevant is this to the end admin? I'm not familiar enough to know if this is important to call out or if it could be dropped.

> - A writable `emptyDir` is now mounted at `/mnt/load-balancer` when load balancing is active (`replicas > 1` or `loadBalancer.forceEnabled: true`), so the prestart script can write the load-balancer config when the pod runs unprivileged (`pod.runAsRoot: false`).

Reviewer: ```suggestion
- A writable `emptyDir` is now mounted at `/mnt/load-balancer` when load balancing is active (`replicas > 1` or `loadBalancer.forceEnabled: true`). The prestart script writes the load-balancer config in this directory when the pod runs unprivileged (`pod.runAsRoot: false`).
```

(Or perhaps "can write" instead of "writes"; I don't know the details.)

- In root mode (`pod.runAsRoot: true`), the pod now sets `fsGroup` (and `fsGroupChangePolicy: OnRootMismatch`) to `serviceAccountUserId` when `launcher.enabled` is true. The Kubernetes launcher runs as the non-root `serviceAccountUser` even in a root pod; without a group-readable ServiceAccount token it failed to start (`Permission denied` reading the token). Non-launcher root deployments are unaffected.
- In rootless mode (`pod.runAsRoot: false`), the chart-managed `secure-cookie-key` is now projected into `/mnt/secret-configmap/rstudio/` even when it is autogenerated. The rootless `launcher.conf` default points `secure-cookie-key-file` there so the launcher shares rserver's cookie key; previously the file was unmounted with the default key and the launcher read a missing file.
- The launcher prestart script now creates its scratch dirs (`/var/lib/rstudio-launcher/Local`, `/Kubernetes`) setgid and group-writable, so the launcher (which runs as the non-root `serviceAccountUser`) can write to them in root-mode pods without a privileged ownership change.

- Workbench pods can now run as a non-root user. Set `pod.runAsRoot: false` (new value, default `true`) to run unprivileged.

> - When running unprivelaged SCIM must be used for user management.

Reviewer 2: I still don't fully understand how SCIM works here but JIT doesn't. Fundamentally it seems like the requirement is that Workbench native user provisioning must be used, not necessarily the exact provisioning method?

Author: Ah! Yeah I'm conflating the two here. I'll be clearer. This work with SCIM as well as JIT. You need to use openid/saml or some other auth provider - local auth basically doesn't work because that leans on pam.
